### DISCUSSION TITLE
Persistent 422 Error with FastApi with an endpoint that accepts parameters and a file upload

### DISCUSSION THREAD

**Discussion by markjrouse** [fastapi/fastapi#10652] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uuid
from pydantic import BaseModel
from fastapi import APIRouter, status, File, UploadFile, Depends, FastAPI
from fastapi.responses import JSONResponse, FileResponse
from starlette.background import BackgroundTask
import pytest
from httpx import AsyncClient

from core.responses import APIResponse
from core.config import settings
from models.pcaps import Pcaps
from models.pcapcreate import PcapCreate


class PcapCreate(BaseModel):
    """
    API schema for posting a new file.
    """

    test_run_id: uuid.UUID
    component_id: uuid.UUID
    pcap_create_user: uuid.UUID


@router.post(
    "/",
    summary="upload a new pcap file",
    description="Uploads a new pcap file.",
    response_model=ACK,
    name="pcaps:create-pcap",
    status_code=status.HTTP_201_CREATED,
    responses={404: {"model": NACK}},
)
async def upload_pcap(
        pcap: PcapCreate = Depends(),
        file: UploadFile = File(...),
) -> JSONResponse:
    """
    Endpoint view for uploading a new file.
    """
    instance = await Pcaps(**pcap.dict(exclude_unset=True)).save()

    pcap_id = str(instance.pcap_id)

    success_text: str = "success"
    ack_message: dict = APIResponse.ack(pcap_id, success_text)

    return JSONResponse(
        status_code=status.HTTP_201_CREATED, content=ack_message
    )

async def test_endpoint_create_valid_pcap_201(
    app: FastAPI,
    client: AsyncClient,
) -> None:
    """
    Tests if a POST request with a file upload returns a 201 response.
    """
    files = {
        'file': open('tests/test.pcap', 'rb'),
    }
    params = {
        "test_run_id": "e3db3e6d-f3db-4d9a-8db5-9ee3e3396ff9",
        "component_id": "d060307d-8b84-44b4-848d-4bd910219d92",
        "pcap_create_user": "6f3300ac-d478-4aa1-8316-c6f30b8402ce",
    }

    response = await client.post(
        app.url_path_for("pcaps:create-pcap"),
        params=params,
        files=files
    )
    print(response.url)
    print(response.text)
    assert response.status_code == status.HTTP_201_CREATED
```

### Description

I have an endpoint defined for uploading a file, and accepting some parameters.  Through the swagger interface this endpoint works perfectly.  My 3 parameters are saved to my database, and my file is uploaded.  However, creating a test, to try and replicate posting to this endpoint a file, and 3 parameters, results in FastAPI returning a 422 error with {"detail":[{"loc":["body","file"],"msg":"field required","type":"value_error.missing"}]}.  I'm guessing I'm doing something wrong, but can't figure where I'm going wrong.

### Operating System

macOS

### FastAPI Version

0.103.1

### Pydantic Version

1.10.8

### Python Version

3.11.6

**Comment by jgould22** [comment]
Try changing your request to look more like the multipart upload described on this page.

https://www.python-httpx.org/quickstart/

If that doesnt work copy the curl example from the swagger doc and paste it here so we can see what the request that works looks like.

**Reply by markjrouse** [reply]
I have tried that as well.  It doesn't work for me - same error.  Here is the curl output from the swagger interface that works:

```
curl -X 'POST' \
  'http://127.0.0.1:8000/api/rma/pcaps/?test_run_id=e3db3e6d-f3db-4d9a-8db5-9ee3e3396ff9&component_id=d060307d-8b84-44b4-848d-4bd910219d92&pcap_create_user=6f3300ac-d478-4aa1-8316-c6f30b8402ce' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'file=@test.pcap'
```

**Reply by markjrouse** [reply]
I've tried this as an example, but same error:

`files = {'file': ('tests/test.pcap', open('tests/test.pcap', 'rb'), 'application/octet-stream')}`

**Reply by JavierSanchezCastro** [reply, marked as answer]
Hi, provide a more basic example so that we can replicate your error.

**Comment by 70omer** [comment]
HI,
@markjrouse did you found the solution to your question?
I'm stuck on the same problem and could really use some help.

Thanks!

**Reply by 70omer** [reply]
Hi,
I changed the code from:
`files = {'file': ('tests/test.pcap', open('tests/test.pcap', 'rb'), 'application/octet-stream')}`
to:
`files = [('file', ('tests/test.pcap', open('tests/test.pcap', 'rb'), 'application/octet-stream'))]`
And it worked!

When I was digging in to the UploadFile in FastAPI, I noticed two important requirements that need to be fulfilled:
1. A FastAPI endpoint (in this case, file) 
2. The value should be a tuple containing the file path & the binary file.
